### ISSUE TITLE
getting "error TS2677: A type predicate's type must be assignable to its parameter's type."

### ISSUE THREAD

**Issue by Heinan-Phau** [typegoose/typegoose#772]
## Versions
- System: linux
- NodeJS: 14.18.0
- Typescript: 4.9.3
- Compiler / Transpiler: tsc
- Typegoose(NPM): ^9.13.0
- Typegoose(GIT): commithash
- mongoose: 6.7.3

## What is the Problem?
`npm run tsc` fails to build with this error:
```

node_modules/@typegoose/typegoose/lib/typeguards.d.ts:8:82 - error TS2677: A type predicate's type must be assignable to its parameter's type.
  Type 'T extends DocumentType<infer T1, infer T2> ? DocumentType<T1, T2> : T extends object ? DocumentType<T, BeAnObject> : never' is not assignable to type 'Ref<T, S>'.
    Type 'DocumentType<unknown, unknown> | (T extends object ? DocumentType<T, BeAnObject> : never)' is not assignable to type 'Ref<T, S>'.
      Type 'DocumentType<unknown, unknown>' is not assignable to type 'Ref<T, S>'.
        Type 'DocumentType<unknown, unknown>' is not assignable to type 'S'.
          'S' could be instantiated with an arbitrary type which could be unrelated to 'DocumentType<unknown, unknown>'.

8 export declare function isDocument<T, S extends RefType>(doc: Ref<T, S>): doc is T extends DocumentType<infer T1, infer T2> ? DocumentType<T1, T2> : T extends object ? DocumentType<T> : never;
```
## Do you know *why* it happens?

*no*

---

**Comment by hasezoey** [comment]
can reproduce, problem is because of using typescript `4.9`, for now `4.8` works - will try to put out a fix

**Comment by Heinan-Phau** [comment]
thanks for the quick response. please alert me when it is fixed if it is not too much to ask :)

**Comment by hasezoey** [comment]
i dont know if i can fix this anytime soon, see https://github.com/microsoft/TypeScript/issues/51638

**Comment by hasezoey** [comment]
i found a workaround for now, will be in typegoose 9.13.1 (funnily this workaround suppresses the error while still doing its job)

**Comment by Jeremiah-Ropo** [comment]
I'm having the same bug here, and my tsc version is 4.8.4, what should I do?

**Comment by hasezoey** [comment]
> I'm having the same bug here, and my tsc version is 4.8.4, what should I do?

are you maybe using VSCode? when yes, you may not have it set to use the project's typescript version and so are using 4.9

aside from that, a fix (hotfix) has already been pushed and released as 9.13.1

---

i still keep this issue open until the "full" fix (which is a refactor of the type) is released (which is 4b3520e55b8b13947cce2a28d83b53825e22fd60, which will be in 10.0)

**Comment by hasezoey** [comment]
i just noticed that the hotfix works as long as the source-file is directly used, but the declaration file (compiled output) does not work because it did not actually transfer the `@ts-ignore`

so the only workaround is to use a older typescript version (being 4.8) and waiting for typegoose 10.0 which updated those types to work on typescript 4.9

**Comment by hasezoey** [comment]
The Typescript issue i had opened has been closed as "not planned" (and more or less "working as intended"), so there will be no fix for typescript 4.9 for typegoose 9.x, see https://github.com/microsoft/TypeScript/issues/51638

Note: typegoose 10.0 will have typescript 4.9 support (and minimal requirement) and a beta has already been released
